linter: warn if expect_num_outputs is missing

bernt-matthias opened the pull request:
if outputs have filters or are discovered

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
linter: warn if expect_num_outputs is missing

if outputs have filters or are discovered

mvdbeek reviewed (commented):
Can you push this into a profile version ?
I don't think this should be a warning for old tools.

mvdbeek commented on lib/galaxy/tool_util/linters/tests.py:
```diff
@@ -26,6 +26,12 @@ def lint_tsts(tool_xml, lint_ctx):
+        discover_datasets = tool_xml.findall("./outputs//discover_datasets")
+        filter = tool_xml.findall("./outputs//filter")
+        if len(discover_datasets) + len(filter) > 0 and "expect_num_outputs" not in test.attrib:
+            lint_ctx.warn("Test should specify 'expect_num_outputs' if outputs have filters or use discover_datasets")
```
There was a discussion about this, maybe we should first do some consistent counting of collections and datasets and describe that.

bernt-matthias replied:
Adapted it to the latest state of https://github.com/galaxyproject/galaxy/pull/7894

bernt-matthias committed:
linter warn for missing expect_num_outputs only if filters are present

discovered datasets are not counted. see:

https://github.com/galaxyproject/galaxy/pull/7894
https://github.com/galaxyproject/galaxy/pull/7894/commits/2aa6831c740126bcd3b6e10e9a0d282aca3b4ba7

bernt-matthias committed:
Merge branch 'dev' into topic/lint-expect-num-outputs

bernt-matthias committed:
post merge cleanup

jmchilton commented on lib/galaxy/tool_util/linters/tests.py:
```diff
@@ -38,11 +38,16 @@ def lint_tsts(tool_xml, lint_ctx):
                 continue
             if len(assertions) > 1:
-                lint_ctx.error(f"Test {test_idx}: More than one {ta} found. Only the first is considered.")
+                lint_ctx.error(f"Test {test_idx}: More than one {ta} found. Only the first is considered.", node=test)
```
This should be a warn, but unrelated change.

gallardoalba replied:
I think it does not allow to include of more than one assertion per test, which I'm not sure if it is really practical.

bernt-matthias replied:
```suggestion
                lint_ctx.warn(f"Test {test_idx}: More than one {ta} found. Only the first is considered.", node=test)
```

bernt-matthias replied:
Thanks for the suggestion @jmchilton .. should I apply it in a separate PR?

@gallardoalba .. its that you can have only one `assert_command` (`assert_stderr`, or `assert_stdout`) per test. Within these you can have an arbitrary number of assertions.

gallardoalba replied:
Ahhh, I understand. Thanks a lot!
